### ISSUE TITLE
Enable nullable?

### ISSUE THREAD

**Issue by Happypig375** [LayoutFarm/Typography#191]
If you accept incorporating nullable reference types then I will start working on an annotation PR 😃

**Comment by zwcloud** [comment]
No. I don't think current Typography's API structures and code quality are good enough to be refactored to enable nullable.

Typography should be refactored thoroughly to clean up APIs first, then start migrate to nullable reference. 
Some points:
* public method, properties and temp structures everywhere: reduce public matters
* so many middle temp classes and structs
* ambiguous naming: Glyph, GlyphPlan, GlyphSetPosition, GlyphPointF
* missing unit test: Typography's usefulness is proven through large numbers of visual demos but the lack of unit test is quite painful when one want to refactor: go through visual demos is not a efficient and human-friendly way to test if the refactor modification breaks anything.

**Comment by Happypig375** [comment]
I would say that code quality is independent of nullable annotations.

Nullable annotations require little to no refactors. It is just adding `?`s after the types, which just expresses intent of whether nulls are accepted. Therefore, any problems arising from refactoring do not exist.

**Comment by zwcloud** [comment]
@Happypig375  Emm.. I think you can just have a try. There will be `?` and `!` everywhere and makes nullable reference meaningless, and in turn makes the code not readable.

**Comment by Happypig375** [comment]
Normally `!` won't be required but I'll try.

**Comment by prepare** [comment]
Thank you @Happypig375 ,

I agree with @zwcloud.

The library needs some (many) arrangements.

**Comment by charlesroddie** [comment]
> @zwcloud There will be ? and ! everywhere

@Happypig375 when you implemented this what did you find? How much is safety improved, or is it that only the unsafe places have been identified? Did everything get ?s and !s or was the compiler identifying a lot of safe paths?

@zwcloud @prepare do you still maintain that other cleanup tasks have to be done first? I don't see why that argument holds; these seem independent. Do you have a suggested path to get NRTs in? Is the difficulty the need to test, or that the changes have been too extensive to review?

The official .Net [guidance](https://devblogs.microsoft.com/dotnet/embracing-nullable-reference-types/) is that NRTs should become ubiquitous with in the .Net 5 timeframe, i.e. by Nov, but of course there are parts of .Net that have historically been slow to update.

**Comment by prepare** [comment]
> @prepare do you still maintain that other cleanup tasks have to be done first?

Yes, This week task=> Typography.TextServices is under development and refactoring.

and I need to test with these too (https://github.com/LayoutFarm/Typography/pull/192#issuecomment-635820167)

**Comment by Happypig375** [comment]
>when you implemented this what did you find?

Legacy-style projects (non-SDK projects) do not support nullable reference types: https://github.com/dotnet/project-system/issues/5551

I had to include refactors to merge the .NET Framework 2.0 projects with the .NET Standard 2.0 projects into multi-targeting SDK-style projects.
>How much is safety improved, or is it that only the unsafe places have been identified?

Yes. However, initialization subroutines called by constructors which initializes fields to non-nullable states were tracked, leading to the constructor not being recognized as initializing all non-nullable fields. This can be directly solved by `MemberNotNullAttribute` as specified in https://github.com/dotnet/csharplang/issues/3297 in the future, but modifying these initialization subroutines to be functions instead, and performing assignment in the constructor just makes the logic more clear.
>Did everything get ?s and !s or was the compiler identifying a lot of safe paths?

I avoided the use of `!` as much as possible because each `!` represents a loss of nullable safety. In cases where the nullability of `var`s are inferred wrongly, spelling out the type usually solves the issue.

**Comment by zwcloud** [comment]
`nullable` just conflicts with my project's configuration, so I cannot include Typography's source code inside the project directly.

**Comment by Happypig375** [comment]
Just add another project with nullable enabled containing Typography and reference it by your existing project.

**Comment by charlesroddie** [comment]
> I had to include refactors to merge the .NET Framework 2.0 projects with the .NET Standard 2.0 projects into multi-targeting SDK-style projects.

@prepare I see that not all projects here have been updated to sdk/netstandard2.0. Is the change to netstandard still in progress?

**Comment by prepare** [comment]
Yes, not all projects have been updated to sdk/netstandard2.0

I support .net framework 2.0 and netstandard2.0 projects (https://github.com/LayoutFarm/Typography/tree/master/Build).

Both set have active maintenance.

**Comment by charlesroddie** [comment]
> I support .net framework 2.0 and netstandard2.0 projects

That is crazy!
